File: backend/app/api/favorites.py
from fastapi import APIRouter, Depends, HTTPException, status
from motor.motor_asyncio import AsyncIOMotorDatabase
from typing import List, Dict, Any

from app.core.database import get_database
from app.models.user_behavior import FavoriteRequest, FavoriteResponse, UserBehaviorStats
from app.services.favorite_service import FavoriteService
from app.api.deps import get_current_user
from app.models.user import User

router = APIRouter()

@router.post("/add", response_model=FavoriteResponse)
async def add_favorite(
    request: FavoriteRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """
    添加收藏文章
    """
    favorite_service = FavoriteService(db)
    result = await favorite_service.add_favorite(current_user.id, request.content_id)
    
    if not result.success:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=result.message
        )
    
    return result

@router.delete("/remove/{content_id}", response_model=FavoriteResponse)
async def remove_favorite(
    content_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """
    取消收藏文章
    """
    favorite_service = FavoriteService(db)
    result = await favorite_service.remove_favorite(current_user.id, content_id)
    
    if not result.success:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=result.message
        )
    
    return result

@router.get("/list", response_model=List[Dict[str, Any]])
async def get_user_favorites(
    limit: int = 50,
    current_user: User = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """
    获取用户收藏的文章列表
    """
    favorite_service = FavoriteService(db)
    favorites = await favorite_service.get_user_favorites(current_user.id, limit)
    return favorites

@router.get("/search", response_model=List[Dict[str, Any]])
async def search_user_favorites(
    query: str = "",
    limit: int = 50,
    current_user: User = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """
    搜索用户收藏的文章
    """
    favorite_service = FavoriteService(db)
    favorites = await favorite_service.search_user_favorites(current_user.id, query, limit)
    return favorites

@router.get("/check/{content_id}")
async def check_favorite_status(
    content_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """
    检查文章是否已收藏
    """
    favorite_service = FavoriteService(db)
    is_favorited = await favorite_service.is_favorited(current_user.id, content_id)
    return {"is_favorited": is_favorited}

@router.get("/stats", response_model=UserBehaviorStats)
async def get_user_behavior_stats(
    current_user: User = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """
    获取用户收藏行为统计
    """
    favorite_service = FavoriteService(db)
    stats = await favorite_service.get_user_behavior_stats(current_user.id)
    return stats

@router.get("/count")
async def get_user_favorites_count(
    current_user: User = Depends(get_current_user),
    db: AsyncIOMotorDatabase = Depends(get_database)
):
    """
    获取用户收藏总数
    """
    favorite_service = FavoriteService(db)
    count = await favorite_service.get_user_favorites_count(current_user.id)
    return {"count": count} 

File: frontend-vue/src/api/favorites.ts
import api from './request'

const API_BASE_URL = '/api/v1'

export interface FavoriteRequest {
  content_id: string
}

export interface FavoriteResponse {
  success: boolean
  message: string
  learned_tags?: {
    energy_types: string[]
    regions: string[]
    business_fields: string[]
  }
  total_favorites: number
}

export interface FavoriteItem {
  _id: string
  content_id: string
  favorited_at: string
  title: string
  publish_date: string
  source: string
  type: string
  basic_info_tags: string[]
  region_tags: string[]
  energy_type_tags: string[]
  business_field_tags: string[]
  beneficiary_tags: string[]
  policy_measure_tags: string[]
  importance_tags: string[]
  link?: string
}

export interface UserBehaviorStats {
  user_id: string
  total_favorites: number
  energy_type_interests: Record<string, number>
  region_interests: Record<string, number>
  last_activity?: string
}

class FavoritesAPI {
  // 添加收藏
  async addFavorite(contentId: string): Promise<FavoriteResponse> {
    const response = await api.post(`/favorites/add`, {
      content_id: contentId
    })
    return response.data
  }

  // 取消收藏
  async removeFavorite(contentId: string): Promise<FavoriteResponse> {
    const response = await api.delete(`/favorites/remove/${contentId}`)
    return response.data
  }

  // 获取收藏列表
  async getFavoritesList(limit: number = 50): Promise<FavoriteItem[]> {
    const response = await api.get(`/favorites/list`, {
      params: { limit }
    })
    return response.data
  }

  // 搜索收藏文章
  async searchFavorites(query: string, limit: number = 50): Promise<FavoriteItem[]> {
    const response = await api.get(`/favorites/search`, {
      params: { query, limit }
    })
    return response.data
  }

  // 检查收藏状态
  async checkFavoriteStatus(contentId: string): Promise<boolean> {
    const response = await api.get(`/favorites/check/${contentId}`)
    return response.data.is_favorited
  }

  // 获取收藏总数
  async getFavoritesCount(): Promise<number> {
    const response = await api.get(`/favorites/count`)
    return response.data.count
  }

  // 获取用户行为统计
  async getUserBehaviorStats(): Promise<UserBehaviorStats> {
    const response = await api.get(`/favorites/stats`)
    return response.data
  }
}

export const favoritesAPI = new FavoritesAPI() 